src/cargo/core/compiler/job_queue.rs

```diff
@@ -27,22 +78,70 @@ use crate::util::{internal, profile, CargoResult, CargoResultExt, ProcessBuilder
 use crate::util::{Config, DependencyQueue};
 use crate::util::{Progress, ProgressStyle};
 
-/// A management structure of the entire dependency graph to compile.
-///
+/// This structure is backed by the `DependencyQueue` type and manages the
+/// queueing of compilation steps for each package. Packages enqueue units of
+/// work and then later on the entire graph is converted to DrainState and
+/// executed.
+pub struct JobQueue<'a, 'cfg> {
+    queue: DependencyQueue<Unit<'a>, Artifact, Job>,
+    counts: HashMap<PackageId, usize>,
+    timings: Timings<'a, 'cfg>,
+}
+
 /// This structure is backed by the `DependencyQueue` type and manages the
 /// actual compilation step of each package. Packages enqueue units of work and
 /// then later on the entire graph is processed and compiled.
-pub struct JobQueue<'a, 'cfg> {
+///
+/// It is created from JobQueue when we have fully assembled the crate graph
+/// (i.e., all package dependencies are known).
+struct DrainState<'a, 'cfg> {
     queue: DependencyQueue<Unit<'a>, Artifact, Job>,
     tx: Sender<Message>,
     rx: Receiver<Message>,
-    active: HashMap<u32, Unit<'a>>,
+    active: HashMap<JobId, Unit<'a>>,
     compiled: HashSet<PackageId>,
     documented: HashSet<PackageId>,
     counts: HashMap<PackageId, usize>,
     progress: Progress<'cfg>,
     next_id: u32,
     timings: Timings<'a, 'cfg>,
+
+    /// Tokens that are currently owned by this Cargo, and may be "associated"
+    /// with a rustc process. They may also be unused, though if so will be
+    /// dropped on the next loop iteration.
+    ///
+    /// Note that the length of this may be zero, but we will still spawn work,
+    /// as we share the implicit token given to this Cargo process with a
+    /// single rustc process.
+    tokens: Vec<Acquired>,
+
+    /// rustc per-thread tokens, when in jobserver-per-rustc mode.
+    rustc_tokens: HashMap<JobId, Vec<Acquired>>,
+
+    /// This represents the list of rustc jobs (processes) and associated
+    /// clients that are interested in receiving a token. Note that each process
+    /// may be present many times (if it has requested multiple tokens).
+    // We use a vec here as we don't want to order randomly which rustc we give
+    // tokens to.
+    to_send_clients: Vec<(JobId, Client)>,
```

Comment:
I think the ideal data structure that we want here is some sort of "priority map", although I'm not sure that this exists. Given my comments below I think we can reduce this to `Vec<JobId>`, but I think that would be best represented as `HashMap<JobId, u32>` instead to make removal more efficient. 

I think this is generally fine for now, though, and we can use `map.iter().next()` to get the next rustc to allocate a token. We can always figure out a better prioritization system later for allocating tokens to rustc's, but I'd like to make sure we're not doing a bunch of accidental O(n^2) work in Cargo here.

Reply:
How would you feel about a `BTreeMap<JobId, Client>`? That means the oldest (i.e., longest running) jobs get the tokens first?

Reply:
I guess given the other refactor suggested this could also just be a `BinaryHeap<JobId>`, right?

Reply:
Oh yeah actually a `BTreeMap` which means oldest sounds pretty reasonable. I don't think we want `BinaryHeap` as-is, (we'd want a custom comparison function at least) and I'm not sure if that supports arbitrary deletion. In any case let's stick to `BTreeMap` for